### ISSUE TITLE
CE_STING doesn't let AFTP know something, so it loops endlessly

### ISSUE THREAD

**Issue by atarijookie** [atarijookie/ce-atari#15]
When you cancel FTP LIST (or similar longer operation) in AFTP, it probably waits for something, and it waits endlessly. 

Either CE_STING doesn't notifiy AFTP correctly somehow that something has happened, or some structure (CIB / CAB) wasn't updated, or something similar.

**Comment by miniupnp** [comment]
I've got aFTP 1.55b, I'm trying to cancel operation

**Comment by miniupnp** [comment]
I cannot get Atarirc to work. Which version could you advise for my TOS 1.06 STE ?

**Comment by atarijookie** [comment]
Well, if you're talking about AtarIRC from Lonny Pursel, that is intended only for newer machines with newer TOS (TT, Falcon, and clones?), so Lonny told me it won't work properly on ST because it lacks something in TOS / AES, and he recommended IRC.TOS - I can't find it now online, so you could get it from here:
https://www.dropbox.com/s/27kkzwfexx0evcf/irc_tos.zip?dl=0
or this one:
https://www.dropbox.com/s/hwssuqdtce5vbeh/irc.zip?dl=0

It's text console only, but I was talking through it on ##Atari like 2 weeks ago.

**Comment by miniupnp** [comment]
thanks, IRC.TOS works well.

I've just tested litchi 1.1j
It does try to connect, but I've got "Authentication failed" in the console along with "connected to 192.168.10.254"
In the log, it looks like 2 connections are open : 

```
handleAcsiCommand: 40 43 45 04 50 00
handleAcsiCommand - CE specific command - module: 04
NET_CMD_RESOLVE
NetAdapter::resolveStart - will resolve: 192.168.10.254

handleAcsiCommand: 40 43 45 04 55 00
handleAcsiCommand - CE specific command - module: 04
NET_CMD_RESOLVE_GET_RESPONSE
NetAdapter::resolveGetResp -- returned 1 IPs

handleAcsiCommand: 40 43 45 04 10 00
handleAcsiCommand - CE specific command - module: 04
NET_CMD_TCP_OPEN
NetAdapter::conOpen() -- remoteHost: 192.168.10.254, remotePort: 21, buff_size: 2000, localPort: 0
NetAdapter::conOpen() -- will connect to host
NetAdapter::conOpen_connect() - non-blocking TCP is connecting, going to TSYN_SENT status
NetAdapter::getLocalPort - getsockname() returned local port: 57448
NetAdapter::conOpen_connect() - returning 81 as handle for slot 1

handleAcsiCommand: 40 43 45 04 47 00
handleAcsiCommand - CE specific command - module: 04

ReadWrapper::getCount() - TCP connection has 0 bytes
NetAdapter::updateCons_active(0) -- bytesInSocket: 0, status: 4
ReadWrapper::getCount() - TCP connection has 36 bytes
NetAdapter::updateCons_active(1) -- bytesInSocket: 36, status: 2
NetAdapter::updateCons_active() -- connection 1 is now TESTABLISH - because no error
NetAdapter::conUpdateInfo [0] - status: 4, localPort: 57447, remoteHost: c0a80afe, remotePort: 21, bytesInSocket: 0
NetAdapter::conUpdateInfo - connection 1 has 36 bytes waiting to be read
NetAdapter::conUpdateInfo [1] - status: 4, localPort: 57448, remoteHost: c0a80afe, remotePort: 21, bytesInSocket: 36
NetAdapter::conUpdateInfo - imcpCnt: 0

```

**Comment by miniupnp** [comment]
I think I have the same bug with litchi that you have with AFTP

**Comment by tin-nl** [comment]
Can you have a test with ftp://ftp.funet.fi/pub/atari/ ? Does this work for you? Which hdd driver do you use and is your ce_sting with or w/o vbl?

**Comment by miniupnp** [comment]
it works a bit better with ftp.funet.fi. it says the authentification succeeded but never get the directory listing.
I'm using the "'without vbl" driver, maybe it matters.

**Comment by miniupnp** [comment]
is litchi working for someone ?

**Comment by tin-nl** [comment]
That's interresting - I was able to dl complete directories from there (9 days ago, see #13). 
Which HDD driver are you using? Maybe that's the difference?

**Comment by miniupnp** [comment]
only CE_DD

**Comment by tin-nl** [comment]
Could you try the same with a native driver and an SD card as write target? Maybe the missing C drive poses a problem here.

**Comment by tin-nl** [comment]
(I was using the non-vbl version, too, btw.)

**Comment by miniupnp** [comment]
I'm not even getting file listing... I'm using Litchi 1.1J with TOS 1.06, what versions are you testing ?

**Comment by tin-nl** [comment]
AFAIK the latest, litchi 1.3f. TOS 1.04.

**Comment by miniupnp** [comment]
OK. lichi 1.3f works, lichi 1.1 doesn't. (I tested both on my TOS 1.06 STE)
according to the log, they are not using the same Sting functions.
lichi 1.1 makes use of CNget_chars()
lichi 1.3 only of CNget_block().
I guess there is a bug in ce_sting implementation of CNget_chars()

**Comment by tin-nl** [comment]
Good find! :)

**Comment by miniupnp** [comment]
@atarijookie does AFTP uses CNget_chars() ?

btw, does anyone know where there is a proper Sting API documentation ? I only have the "DEV-KIT" from Sting with the .h files...

**Comment by atarijookie** [comment]
> I guess there is a bug in ce_sting implementation of CNget_chars()

Might be, but the test for CNget_char() in test_sting prg seems to succeed. Maybe I'm missing something there...

> btw, does anyone know where there is a proper Sting API documentation?

There is this Sting guide, I have one copy here:
[http://joo.kie.sk/cosmosex/sting_docu.zip](http://joo.kie.sk/cosmosex/sting_docu.zip) 

...and you can view it through your browser, by using Opichal's online viewer, e.g. like this:
http://phoenix.inf.upol.cz/~opichals/libhyp/hypview.cgi?url=http%3A%2F%2Fjoo.kie.sk%2Fcosmosex%2Fsting_docu.zip&mask=STING.HYP&index=43

But please note that some behavior is not mentioned there, and some return codes (and thus another behavior) is wrong there and doesn't match with the real return codes of Sting 1.26 - I have implemented the driver according to documentation, then I've run the test with Sting 1.26 and EtherNEC (network card for ST), and then I've fixed the expected results according to real behavior, not documentation - these apps can run fine with the real Sting, and I guess they could fail if everything would be like it's written in that documentation. 

So before fixing things just according to documentation - be sure to test it with real Sting first.

**Comment by miniupnp** [comment]
Thanks a lot !!!

**Comment by miniupnp** [comment]
I have improved my test : a540d7ee6fedda753c3a058f63a08e9c20f88189
It now shows there are "data mismatch" at random times

**Comment by atarijookie** [comment]
Wow, good find and... damn, how does that happen? :-o

**Comment by miniupnp** [comment]
Can someone else test with the new test to see if its not my setup.
It would be great to check if the test run with real sting

**Comment by atarijookie** [comment]
I can check it with my device, and I can (and will) check it with real sting during this night - around 00:00 - 03:00, when I'm doing atari stuff instead of sleeping...

**Comment by miniupnp** [comment]
I added a menu to the test program : afaaa03f9675f3be8ba49ef7fe5479f3580df64c
interestingly, my test (020a) always fails when I run all tests, but not when I run only 02xx tests...

**Comment by atarijookie** [comment]
I'm sorry, I just can get to that commit :-( I'm still having a mess with the git, branches and your fork with branches...

I've done git pull on my ce-atari, the test isn't there because it's probably in your fork / branch... So I did 'git clone' on your ce-atari repo, and the test still isn't there. So I've tried to watch for branches in your and mine repo, and it only shows master, while in the mean time here on the github page I can see that the test is in some tst_sting branch... I will try to do clone with branch specified, and if that fails, then I just need help getting to it...

**Comment by atarijookie** [comment]
...and it went like this:

```
$ git clone -b tst_sting https://github.com/miniupnp/ce-atari.git
Cloning into 'ce-atari'...
fatal: Remote branch tst_sting not found in upstream origin
```

**Comment by miniupnp** [comment]
WTF ?
Don't clone my repository, one local repository is enough !
you should have "remote" repositories set.
For example, on my PC :

```
$ git remote -v
atarijookie     https://github.com/atarijookie/ce-atari.git (fetch)
atarijookie     https://github.com/atarijookie/ce-atari.git (push)
origin  https://miniupnp%40free.fr@github.com/miniupnp/ce-atari.git (fetch)
origin  https://miniupnp%40free.fr@github.com/miniupnp/ce-atari.git (push)
```

on your PC it should be something like

```
origin     https://github.com/atarijookie/ce-atari.git (fetch)
origin     https://github.com/atarijookie/ce-atari.git (push)
```

To add a remote you do
`$ git remote add miniupnp https://github.com/miniupnp/ce-atari.git`
and then
`$ git fetch --all`
It will fetch my branches
just `$ git branch -a` to list all branches (local & remote)
you can also use `$ git log --all --decorate --graph` to have a nice graphical view and check what is in sync, or what need to be pushed :)
then it is just a mater of doing `$ git checkout miniupnp/test_sting` to go to my branch

**Comment by atarijookie** [comment]
> WTF ?

I know, I know - I'm lost and doing weird things while trying to get things done. I'm not used to git, and I still haven't touched anything but like 5 main command... In cases like this I spend too much time figuring out stuff instead of doing coding and testing.

OK, so now I've got your test_sting branch, I'm going to try those things.

**Comment by atarijookie** [comment]
The test_sting fails terribly in that branch, and I guess in master branch it fails terribly, too, like this:
https://dl.dropboxusercontent.com/u/50798228/STNG_TST_new.txt

I've checked out the version under the tag release_ce_main_app_2016-05-01 (dfcc57850a08e4a1ae5e917de7f994d78fd12bad), and the test succeeds with all OK:
https://dl.dropboxusercontent.com/u/50798228/STNG_TST_old.txt

...so this means there's something broken since last release, either in ce_main_app, ce_sting or test_sting, and it seems that we got it merged in the master branch now :) Please can you take a look at it, @miniupnp ?

**Comment by atarijookie** [comment]
The issue seems to come from CE_STING.PRG, as with newer ce_main_app but older ce_sting it works, but with newer ce_sting it doesn't work....

**Comment by miniupnp** [comment]
the only difference I see in cosmosex_fakesting is the VBL handler not installed : 
https://github.com/atarijookie/ce-atari/compare/release_ce_main_app-2016-05-01...master
I'm going to test further tomorrow

**Comment by miniupnp** [comment]
I'm doing complete cross tests and I think I understand what's going on

**Comment by miniupnp** [comment]
many test fails if VBL handler is not installed. It is installed by last tests (ICMP). So if I launch test program twice, it works on 2nd time...
many stuff to fix to have this "without vbl" working properly !

**Comment by miniupnp** [comment]
I also have to fix my test020a !

```
Setup :
1 release_ce_main_app_2016-05-01
  ce_stin1.prg / tst_stn1.prg

2 master
  ce_stin2.prg / tst_stn2.prg

3 test_sting
  ce_stin3.prg / tst_stn3.prg


tests with ce_stin1.prg :
  tst_stn1.prg OK   => STNG_TST_1_1.txt
  tst_stn2.prg OK   => STNG_TST_1_2.txt
  tst_stn3.prg OK   => STNG_TST_1_3.txt
STNG_TST_1_3.txt:[020A] [TCP CNget_char                          ][fail] [03B1] err: CNget_char() data mismatch

tests with ce_stin2.prg:
  tst_stn1.prg FAIL => STNG_TST_2_1.txt
  tst_stn2.prg OK   => STNG_TST_2_2.txt
  tst_stn3.prg OK   => STNG_TST_2_3.txt

 reboot
  tst_stn2.prg FAIL => STNG_TST_2_2_b.txt
  tst_stn3.prg OK   => STNG_TST_2_3_b.txt
  tst_stn1.prg OK   => STNG_TST_2_1_b.txt

 reboot
  tst_stn3.prg ENDLESS LOOP => STNG_TST_2_3_c.txt
  stuck in test [020A], need to add a timeout !

```

**Comment by atarijookie** [comment]
> many stuff to fix to have this "without vbl" working properly !

Well, the original idea was to utilize the VBL for the connection status update, so without the VBL it's more like the remake to fit the idea without VBL ;)

Looking forward to any improvement there will be, keep up the good work! :+1:

**Comment by miniupnp** [comment]
I think everything is related to bug #55

They are timing differences with vbl / without. So the bug appears depending on the time the functions are called.
Clearly that ICMPWrapper is receiving stuff from other sockets ! I'm fixing that and we'll test more later :)

**Comment by miniupnp** [comment]
with my fix for #55, not all tests pass, but the 010x are OK.
It was really strange ti worked for UDP but not TCP.
I know understand the IcmpWrapper somehow "stole" data from TCP socket, but not the other one

**Comment by miniupnp** [comment]
@atarijookie I've got a great improvement in litchi 1.1 working with this patch : fc137d8280cfced70050c990d433b6a5c84ca723
Is it making sense to you ?

**Comment by atarijookie** [comment]
Yes, it makes sense... In case there isn't any data, the nc->readWrapper.peekBlock() could return -1 -- inside of peekBlock() the tcpBytesWaiting() shouldn't anything less than 0, but there isn't a check if the:
totalCount = tcpBytesWaiting();
didn't return 0 as totalCount, so then the:
res  = recv(fd, tmpBuffer, size, MSG_DONTWAIT | MSG_PEEK);
could return that -1 in this case, where no data is available. So if possible, please add:
if(totalCount > 0) { do the recv() } else { skip recv() } inside that readWrapper.peekBlock() to have it fixed also at the right place... 

Then sending 32 bytes isn't that necessary, I guess 16 should be enough (DMA chip transfers them into RAM by 16 bytes blocks), but I wonder if that's necessary at all (but if it shows some improvement, then why the hell not).

**Comment by miniupnp** [comment]
@atarijookie : is this bug really fixed ?
(it was automatically closed by my commit msg which stated "see if that fixes #15 ;)

**Comment by atarijookie** [comment]
@miniupnp I didn't see the AFTP loop endlessly in the last time, but after closing the connection the AFTP doesn't react on mouse clicks, it might be something related to it requiring (and using) something from newer AES.
